addons/portal/controllers/portal.py

```diff
@@ -344,44 +428,61 @@ def attachment_remove(self, attachment_id, access_token=None):
 
         # Validation
-        for field_name in self.MANDATORY_BILLING_FIELDS:
+        for field_name in self.MANDATORY_PARTNER_FIELDS:
             if not data.get(field_name):
-                error[field_name] = 'missing'
+                error[self.MANDATORY_PARTNER_FIELDS[field_name]] = 'missing'
 
         # email validation
         if data.get('email') and not tools.single_email_re.match(data.get('email')):
             error["email"] = 'error'
             error_message.append(_('Invalid Email! Please enter a valid email address.'))
 
+        # country_id and state_id validation and formatting
+        for field in set(['country_id', 'state_id']) & set(data.keys()):
+            try:
+                data[field] = int(data[field]) if data[field] else False
+            except ValueError:
+                data[field] = False
+                error[field] = 'error'
+
+        # type validation
+        if not validate_main_address and data.get('type') and not data['type'] in set(['delivery', 'invoice']):
+            error['type'] = 'error'
+
         # vat validation
         partner = request.env.user.partner_id
-        if data.get("vat") and partner and partner.vat != data.get("vat"):
+        if validate_main_address and data.get("vat") and partner and partner.vat != data.get("vat"):
             if partner.can_edit_vat():
                 if hasattr(partner, "check_vat"):
                     if data.get("country_id"):
-                        data["vat"] = request.env["res.partner"].fix_eu_vat_number(int(data.get("country_id")), data.get("vat"))
+                        data["vat"] = request.env["res.partner"].fix_eu_vat_number(data.get("country_id"), data.get("vat"))
                     partner_dummy = partner.new({
                         'vat': data['vat'],
-                        'country_id': (int(data['country_id'])
-                                       if data.get('country_id') else False),
+                        'country_id': data['country_id'],
                     })
                     try:
                         partner_dummy.check_vat()
-                    except ValidationError:
+                    except ValidationError as err:
                         error["vat"] = 'error'
+                        error_message.append(err.message)
             else:
+                error["vat"] = 'error'
                 error_message.append(_('Changing VAT number is not allowed once document(s) have been issued for your account. Please contact us directly for this operation.'))
 
         # error message for empty required fields
         if [err for err in error.values() if err == 'missing']:
             error_message.append(_('Some required fields are empty.'))
 
-        unknown = [k for k in data if k not in self.MANDATORY_BILLING_FIELDS + self.OPTIONAL_BILLING_FIELDS]
+        unknown = [k for k in data if k not in chain(self.MANDATORY_PARTNER_FIELDS.keys(), self.OPTIONAL_PARTNER_FIELDS.keys())]
```

Comment:
is there any changes here ? (I don't know chain so :shrug: )

Reply:
Keys should be an iterable, isn't "+" working ?

Reply:
keys() returns a `dict_keys` instance and while it is an iterable, the operator + is not supported. chain will return an iterator going through all iterables parameters sequentially. I think it is marginally better to do this than 
```python
{**self.dict1, **self.dict2}.keys()
```

Reply:
And why not casting directly like ``list(self.MANDATORY_PARTNER_FIELDS.keys()) + list(self.OPTIONAL_PARTNER_FIELDS.keys())`` ? Should work and is more easy-to-read python code :)

Reply:
I don't see the problem with chain, it is already used in the odoo codebase, the doc is pretty explanatory when hovering it in an editor, it avoids creating 3 lists and thus is justified in this case imo.